internal/http/services/owncloud/ocdav/tpc.go

```diff
@@ -0,0 +1,415 @@
+			code:    SabredavBadRequest,
+			message: m,
+		})
+		HandleWebdavError(&sublog, w, b, err)
+		return
+	}
+
+	sublog.Debug().Str("overwrite", overwrite).Msg("TPC Push")
+
+	// get Gateway client
+	client, err := s.getClient()
+	if err != nil {
+		sublog.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	ref := &provider.Reference{Path: src}
+	srcStatReq := &provider.StatRequest{Ref: ref}
+	srcStatRes, err := client.Stat(ctx, srcStatReq)
+	if err != nil {
+		sublog.Error().Err(err).Msg("error sending grpc stat request")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if srcStatRes.Status.Code != rpc.Code_CODE_OK && srcStatRes.Status.Code != rpc.Code_CODE_NOT_FOUND {
+		HandleErrorStatus(&sublog, w, srcStatRes.Status)
+		return
+	}
+	if srcStatRes.Info.Type == provider.ResourceType_RESOURCE_TYPE_CONTAINER {
+		sublog.Error().Msg("Third-Party copy of a folder is not supported")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	err = s.performHTTPPush(ctx, client, r, w, srcStatRes.Info, ns)
+	if err != nil {
+		sublog.Error().Err(err).Msg("error performing TPC Push")
+		return
+	}
+	fmt.Fprintf(w, "success: Created")
+}
+
+func (s *svc) performHTTPPush(ctx context.Context, client gateway.GatewayAPIClient, r *http.Request, w http.ResponseWriter, srcInfo *provider.ResourceInfo, ns string) error {
+	src := path.Join(ns, r.URL.Path)
+	dst := r.Header.Get("Destination")
+
+	sublog := appctx.GetLogger(ctx)
+	sublog.Debug().Str("src", src).Str("dst", dst).Msg("Performing HTTP Push")
+
+	// get download url
+	dReq := &provider.InitiateFileDownloadRequest{
+		Ref: &provider.Reference{Path: src},
+	}
+
+	dRes, err := client.InitiateFileDownload(ctx, dReq)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return err
+	}
+
+	if dRes.Status.Code != rpc.Code_CODE_OK {
+		w.WriteHeader(http.StatusInternalServerError)
+		return fmt.Errorf("status code %d", dRes.Status.Code)
+	}
+
+	var downloadEP, downloadToken string
+	for _, p := range dRes.Protocols {
+		if p.Protocol == "simple" {
+			downloadEP, downloadToken = p.DownloadEndpoint, p.Token
+		}
+	}
+
+	// do download
+	httpDownloadReq, err := rhttp.NewRequest(ctx, "GET", downloadEP, nil)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return err
+	}
+	httpDownloadReq.Header.Set(datagateway.TokenTransportHeader, downloadToken)
+
+	httpDownloadRes, err := s.client.Do(httpDownloadReq)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return err
+	}
+	defer httpDownloadRes.Body.Close()
+	if httpDownloadRes.StatusCode != http.StatusOK {
+		w.WriteHeader(httpDownloadRes.StatusCode)
+		return fmt.Errorf("Remote PUT returned status code %d", httpDownloadRes.StatusCode)
+	}
+
+	// send performance markers periodically every PerfMarkerResponseTime (5 seconds unless configured)
+	w.WriteHeader(http.StatusAccepted)
+	wc := WriteCounter{0, time.Now(), w}
```

Comment:
You're never really sending the performance marker, are you?

Reply:
Yes, they are: the GFAL client understands the response as a kind of progressive HTML. Why do you think they are never sent?

Actually, for reference this is a typical output of a `gfal-copy` operation. Though the Davix layer seems to ignore them, FTS takes them into account:

```
Executing TPC in pull mode:
gfal-copy -v --copy-mode pull -D"BEARER:TOKEN=xxxxx" cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc.txt cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     [gfal_module_load] plugin /usr/lib64/gfal2-plugins//libgfal_plugin_file.so loaded with success 
INFO     Davix: Unable to find a proxy or cert/key pair using either X509_USER_* variables or /tmp/x509up_u0
INFO     Davix: Unable to find a proxy or cert/key pair using either X509_USER_* variables or /tmp/x509up_u0
INFO     [gfal_module_load] plugin /usr/lib64/gfal2-plugins//libgfal_plugin_http.so loaded with success 
INFO     [gfal_module_load] plugin /usr/lib64/gfal2-plugins//libgfal_plugin_sftp.so loaded with success 
INFO     Davix: Negative result for operation: HTTP 404 : File not found . After 1 retry
INFO     Davix: Could not execute operation on cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV, error Result HTTP 404 : File not found  after 1 attempts
INFO     Davix: Try to Recover with Metalink...
INFO     Davix: Impossible to Recover with Metalink: Server does not support Metalink standard
INFO     Davix: Negative result for operation: HTTP 404 : File not found . After 1 retry
INFO     Davix: Could not execute operation on cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV, error Result HTTP 404 : File not found  after 1 attempts
INFO     Davix: Try to Recover with Metalink...
INFO     Davix: Impossible to Recover with Metalink: Server does not support Metalink standard
Copying 2110285696 bytes cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
event: [1644566905599] BOTH   GFAL2:CORE:COPY	LIST:ENTER	
INFO     Event triggered: BOTH GFAL2:CORE:COPY LIST:ENTER 
event: [1644566905599] BOTH   GFAL2:CORE:COPY	LIST:ITEM	cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Event triggered: BOTH GFAL2:CORE:COPY LIST:ITEM cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
event: [1644566905600] BOTH   GFAL2:CORE:COPY	LIST:EXIT	
INFO     Event triggered: BOTH GFAL2:CORE:COPY LIST:EXIT 
event: [1644566905600] BOTH   http_plugin	PREPARE:ENTER	cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Event triggered: BOTH http_plugin PREPARE:ENTER cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Davix: Negative result for operation: HTTP 404 : File not found . After 1 retry
INFO     Davix: Could not execute operation on cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV, error Result HTTP 404 : File not found  after 1 attempts
INFO     Davix: Try to Recover with Metalink...
INFO     Davix: Impossible to Recover with Metalink: Server does not support Metalink standard
event: [1644566906306] BOTH   http_plugin	PREPARE:EXIT	cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Event triggered: BOTH http_plugin PREPARE:EXIT cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
event: [1644566906306] BOTH   http_plugin	TRANSFER:ENTER	cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Event triggered: BOTH http_plugin TRANSFER:ENTER cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Trying copying with mode 3rd pull
event: [1644566906306] BOTH   http_plugin	TRANSFER:TYPE	3rd pull
INFO     Event triggered: BOTH http_plugin TRANSFER:TYPE 3rd pull
INFO     Performing a HTTP third party copy
INFO     Davix: Hop: cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Davix: Received: Perf Marker
INFO     Davix: Received: Timestamp: 1644566922
INFO     Davix: Received: Stripe Bytes Transferred: 562821824
INFO     Davix: Received: Strip Index: 0
INFO     Davix: Unknown performance marker, ignoring: Strip Index: 0
INFO     Davix: Received: Total Stripe Count: 1
INFO     Davix: Received: End
INFO     Davix: Received: 
INFO     Davix: Unknown performance marker, ignoring: 
INFO     Davix: Received: Perf Marker
INFO     Davix: Received: Timestamp: 1644566927
INFO     Davix: Received: Stripe Bytes Transferred: 1116601024
INFO     Davix: Received: Strip Index: 0
INFO     Davix: Unknown performance marker, ignoring: Strip Index: 0
INFO     Davix: Received: Total Stripe Count: 1
INFO     Davix: Received: End
monitor: cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV 223320204 110755840 1116601024 5
INFO     Davix: Received: 
INFO     Davix: Unknown performance marker, ignoring: 
INFO     Davix: Received: Perf Marker
INFO     Davix: Received: Timestamp: 1644566932
INFO     Davix: Received: Stripe Bytes Transferred: 1658585087
INFO     Davix: Received: Strip Index: 0
INFO     Davix: Unknown performance marker, ignoring: Strip Index: 0
INFO     Davix: Received: Total Stripe Count: 1
INFO     Davix: Received: End
monitor: cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV 165858508 108396812 1658585087 10
INFO     Davix: Received: 
INFO     Davix: Unknown performance marker, ignoring: 
INFO     Davix: Received: success: Created
INFO     Copy succeeded using mode 3rd pull
event: [1644566950654] BOTH   http_plugin	TRANSFER:EXIT	cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
INFO     Event triggered: BOTH http_plugin TRANSFER:EXIT cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/largefile => cs3s://ruciose.cernbox.cern.ch/remote.php/webdav/eos/project/c/cernbox/test/rucio_se/testtpc-FnyV
```

Reply:
Ah okay, my bad